src/Event/EventFactory.cpp

```diff
@@ -1,4 +1,220 @@
+
+        for (const auto & vehicle : orderedVehicles)
+        {
+            for (const auto & reqt : vehicleReqts)
+            {
+                // TODO - should confirm the Engine/Ladder type
+                if (vehicle->getCurNumCrew() >= reqt->getCurNumCrew()
+                    && vehicle->getCurWaterVolumeLitres() >= reqt->getCurWaterVolumeLitres())
+                {
+                    meetingRequirements.push_back(vehicle);
+                }
+            }
+        }
+        return meetingRequirements;
+    }
+
+    std::vector<Vehicle *> EventFactory::selectVehicles(
+            std::vector<Vehicle *> vehicleReqts, std::vector<Vehicle *> orderedVehicles,
+            BitArray *bitArray) {
+        while (true)
+        {
+            std::vector<Vehicle*> selectedVehicles;
+            uint8_t numVehiclesRequired = vehicleReqts.size();
+            std::vector<Vehicle*> listSubset = vehiclesMeetingRequirements(vehicleReqts, orderedVehicles);
+            float avgWorkFactor = calculateAverageWorkFactor(listSubset);
+
+            for (const auto & vehicle : listSubset)
+            {
+                if (vehicle->getWorkFactor() >= avgWorkFactor)
+                {
+                    if (vehicle->getCurVehicleStatus() == VehicleStatus::Available
+                    || vehicle->getCurVehicleStatus() == VehicleStatus::Returning)
+                    {
+                        selectedVehicles.push_back(vehicle);
+                        --numVehiclesRequired;
+                    }
+                }
+                if (numVehiclesRequired == 0)
+                {
+                    if (modifyVehicleStatus(selectedVehicles, bitArray))
+                    {
+                        return selectedVehicles;
+                    }
+                    else
+                    {
+                        // Break inner for loop and continue outer while loop
+                        // I know, I know, it's a goto - don't be scared
+                        // This is the easiest way to perform the loop iteration needed and won't harm anyone
+                        // despite what everyone says, these can be useful.
+                        goto tryAgain;
+                    }
+                }
+            }
+            for (const auto & vehicle : orderedVehicles) {
+                if (vehicle->getCurVehicleStatus() == VehicleStatus::Available
+                    || vehicle->getCurVehicleStatus() == VehicleStatus::Returning) {
+                    selectedVehicles.push_back(vehicle);
+                    --numVehiclesRequired;
+                }
+                if (numVehiclesRequired == 0)
+                {
+                    if (modifyVehicleStatus(selectedVehicles, bitArray))
+                    {
+                        return selectedVehicles;
+                    }
+                }
+            }
+            // See, it's friendly :)
+            tryAgain:;
```

Comment:
🤔 Seems innocuous enough.

Though I wonder if both loops could be refactored into functions where the first returns a `std::optional< std::vector< Vehicle * > >` and the second function is only invoked if that optional is null?

Reply:
I'll give that a look but it really sounds like you don't want me to use a `goto` :disappointed_relieved: 

This was my one chance :money_mouth_face:

Reply:
Nah, I think this is a good example of where a goto makes sense. Just pointing out an alternative refactoring that avoids it.